### DISCUSSION TITLE
Usage with file based routing

### DISCUSSION THREAD

**Discussion by narduin** [Alexandre-Fernandez/astro-i18n#4] [General]
Hi!

I was wondering if this module could be used with a blog structure, or any kind of dynamic page generation based on `.md` or `.mdx` files.

Let's say I have some static pages like `home` and `about` that I can translate with classic i18n `.json` files.
How would I do for my dynamic pages that astro cerates directly from markdown? Is this module capable of this?

I'm trying to wrap my head around a structure that would work. For example, astro will create pages for all `.md` files inside an `src/pages` directory.

- Should we use that default behaviour or is it not compatible?
- Should we use a folder per language?
- Or maybe we should put our markdown outside of `pages` and render posts using a `[post].astro` file?

I don't really know what's possible and/or what's best. I'll let you know if I figure something out 😄

**Comment by Alexandre-Fernandez** [comment]
At the moment nothing has been done for markdown.
And I'm not sure something needs to be done in that regard. 

If you're importing posts in a page you can just export a `langCode` property from you post's frontmatter and filter the `Astro.glob` results using that and `astroI18n.langCode`. Or you could have a directory for each `langCode` such as `/posts/en` and `/posts/it` and import posts with a string literal `/posts/${astroI18n.langCode}`.

Let's also not forget that `node` should not be used at runtime to prevent compatibility issues (deno, etc).
Considering these points does something really need to be done regarding markdown ? Do you have a meaningful feature in mind ?

**Reply by narduin** [reply]
Right now I have markdown posts outside of `pages` and a `[post].astro` page. I'm actually filtering a frontmatter property against `astroI18n.langCode`, glad to see you're suggesting it as well!

It's working fine but the limitation with this system is that every post must share the same filename, hence the same "post url" like `/posts/first-post` and `/it/posts/first-post` (instead of a translated `first-post` name) in order to work with a language switcher. Or am I missing something maybe?

I tried having a folder per language but coul not wrap my head around having `/posts/en/first-post` and `/it/posts/it/first-post`. I'm not even sure if it's supposed to be like that.

**Reply by Alexandre-Fernandez** [reply]
Assuming you are using another directory than pages, wouldn't importing posts with ```Astro.glob(`/posts/${langCode}/${t("posts.first-post")}.md`)``` work ?
If you want to put your markdown posts inside pages, just put them there, the sync command won't override or delete them.

**Reply by narduin** [reply]
I was not able to use a dynamic import with `Astro.glob`, I'm getting this error:

```
plugin:vite:import-glob] Invalid glob import syntax: Expected glob to be a string, but got dynamic template literal
```

I think it is because of this [in the docs](https://docs.astro.build/en/reference/api-reference/#astroglob).

**Reply by Alexandre-Fernandez** [reply]
Didn't see that, in that case can't you import every post with `*.md` and filter them with JS according to some of their frontmatter attributes ?

**Reply by narduin** [reply]
Yes that's what I'm doing, I just can't translate the name of the posts in the url, which is not a big problem in itself.

> I was not able to use a dynamic import with Astro.glob, I'm getting this error…

Knowing this, do you think a folder for each language inside `pages` is still possible?

**Reply by Alexandre-Fernandez** [reply]
As you said before, [this](https://docs.astro.build/en/guides/markdown-content/#example-dynamic-page-routing) is probably the way to get the url translated.
If you want to have your markdown in `pages`, you can, just put them there, the `i18n:sync` command shouldn't override or delete them.

The real question here is, can `astro-i18n` make this process significantly easier and if so how ?

**Comment by narduin** [comment]
`astro@1.7.0` introduced an experimental feature: [Content collections](https://docs.astro.build/en/guides/content-collections/)

It reminds me of [`nuxt-content`](https://content.nuxtjs.org/) and is an amazing addition to Astro in my opinion.

I'm still playing around with it but it seems to work well with `astro-i18n` through the `getCollection()` function and the `slug` system.
It might even be possible to translate the url with this!
